matchers/collections.go

```diff
@@ -0,0 +1,199 @@
+package matchers
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// Items is a matcher for a slice or array value. It tests that the number of elements is equal to
+// the number of matchers, and that each element matches the corresponding matcher in order.
+//
+//     s := []int{6,2}
+//     matchers.Items(matchers.Equal(6), matchers.Equal(2)).Test(s) // pass
+//     matchers.Items(matchers.Equal(2), matchers.Equal(6)).Test(s) // fail
+func Items(matchers ...Matcher) Matcher {
+	return New(
+		func(value interface{}) bool {
+			elements, err := getSliceOrArrayElementValues(value)
+			if err != nil || len(elements) != len(matchers) {
+				return false
+			}
+			for i, m := range matchers {
+				elementValue := elements[i]
+				if !m.test(elementValue) {
+					return false
+				}
+			}
+			return true
+		},
+		func() string {
+			return "items: " + describeMatchers(matchers, "")
+		},
+		func(value interface{}) string {
+			elements, err := getSliceOrArrayElementValues(value)
+			if err != nil {
+				return err.Error()
+			}
+			if len(elements) != len(matchers) {
+				return fmt.Sprintf("expected slice with %d item(s), got %d item(s)", len(matchers), len(elements))
+			}
+			parts := make([]string, 0, len(matchers))
+			for i, m := range matchers {
+				elementValue := elements[i]
+				if !m.test(elementValue) {
+					parts = append(parts, fmt.Sprintf("item[%d] %s", i, m.describeFailure(elementValue)))
+				}
+			}
+			return strings.Join(parts, ", ")
+		},
+	)
+}
+
+// ItemsInAnyOrder is a matcher for a slice or array value. It tests that the number of elements is
+// equal to the number of matchers, and that each matcher matches an element.
+//
+//     s := []int{6,2}
+//     matchers.ItemsInAnyOrder(matchers.Equal(2), matchers.Equal(6)).Test(s) // pass
+func ItemsInAnyOrder(matchers ...Matcher) Matcher {
```

Comment:
```suggestion
func UnorderedItems(matchers ...Matcher) Matcher {
```

Reply:
Any particular reason? To me they both seem to convey the same thing, and having them both start with "Items" makes them easy to see together in the API docs.

Reply:
I think it looks/feels more declarative, especially if the other was named `OrderedItems` for symmetry.

```golang
matchers.OrderedItems(matchers.Equal(2), matchers.Equal(6)).Test(slice)
matchers.UnorderedItems(matchers.Equal(2), matchers.Equal(6)).Test(slice)

// vs
matchers.Items(matchers.Equal(2), matchers.Equal(6)).Test(slice)
matchers.ItemsInAnyOrder(matchers.Equal(2), matchers.Equal(6)).Test(slice)

//or alternatively

matchers.ItemsOrdered(matchers.Equal(2), matchers.Equal(6)).Test(slice)
matchers.ItemsUnordered(matchers.Equal(2), matchers.Equal(6)).Test(slice)
```

But I suppose it is entirely subjective.

Reply:
It is subjective, and subjectively I don't prefer those alternatives - sorry. To me, "ItemsOrdered" and "ItemsUnordered" sounds like you are making an assertion that "the items are ordered" or "the items are unordered".

And as for the other option - "symmetry" to me is not really as important as discoverability, and since Go docs list methods in alphabetical order, having them both start with the same word makes it clearer that they are two related things. And if anyone sees "Items" and "ItemsInAnyOrder" and is really confused about what the difference is... well, that's what the lengthy doc comments are for.